Test Cases for Live Reload

amolgautam25 opened the pull request:
Adding Test cases for Live Reload.

fosterseth commented on tests/functional/mesh/reload_test.go:
```diff
@@ -0,0 +1,166 @@
+package mesh
```
do we need reload_test.go? If we can keep the testing to just work_test.go that would be ideal

amolgautam25 replied:
I was originally thinking that we can have two different test files : 

`reload_test.go` would just test the reload functionality. 
`work_test.go` would test the work streaming process in case of a reload.

amolgautam25 replied:
I have put all the test in ```work_test.go```

fosterseth commented on tests/functional/mesh/work_test.go:
```diff
@@ -587,6 +607,140 @@ func TestWork(t *testing.T) {
+
+			modifiedData.Nodes["node2"] = &mesh.YamlNode{
+				Connections: map[string]mesh.YamlConnection{},
+				Nodedef: []interface{}{
```
does defining NodeDef here do anything? I thought modifyclimeshfromyaml just ignores NodeDef anyways

amolgautam25 replied:
I have changed this. It will be pushed in the next commit.

fosterseth commented on tests/functional/mesh/work_test.go:
```diff
@@ -587,6 +607,140 @@ func TestWork(t *testing.T) {
+
+			// reload the entire mesh
+			// wait for 20 seconds
+			time.Sleep(20 * time.Second)
```
what are we waiting for exactly? I think we'll want to poll and assert a state rather than sleep an arbitrary time

amolgautam25 replied:
Yeah , correct. This is just me trying to figure out what is wrong.

fosterseth commented on tests/functional/mesh/work_test.go:
```diff
@@ -587,6 +607,140 @@ func TestWork(t *testing.T) {
+			}
+
+			// so that the rsults can start streaming again
+			time.Sleep(10 * time.Second)
```
I've rethought this. Instead of getting the first work results before the mesh is broken, let's get the work results after it's broken. Then when mesh re-establishes, we can check that stdout is larger.

amolgautam25 replied:
okay , i will make that change.

fosterseth commented on tests/functional/lib/mesh/climesh.go:
```diff
@@ -150,6 +150,111 @@ func (m *CLIMesh) Nodes() map[string]Node {
 	return nodes
 }
 
+func populateNodedefConnection(meshDefinition *YamlData, existingMesh *CLIMesh) error {
```
let's rename this to `createNodedefConnections`

let's also keep `connections` plural everywhere, so `NodedefConnections` instead of `NodedefConnection`

amolgautam25 replied:
I have made the required change

fosterseth commented on tests/functional/lib/mesh/climesh.go:
```diff
@@ -374,8 +402,23 @@ func NewCLIMeshFromYaml(meshDefinition YamlData, dirSuffix string) (*CLIMesh, er
 		}
 	}
 
+	for k := range meshDefinition.Nodes {
```
so here we fill in Nodedef, then copy it to nodedefbase, then do nodedef = nodedefbase + nodedefconnections

i think it would be more straightforward to change NewCLIMeshFromYaml to fill in NodedefBase to begin with, then do

nodedef = nodedefbase + nodedefconnections at the end

amolgautam25 replied:
I made the changes

fosterseth commented on tests/functional/lib/mesh/climesh.go:
```diff
@@ -584,6 +675,16 @@ func (m *CLIMesh) WaitForReady(ctx context.Context) error {
 	return nil
 }
 
+// WaitForControlSockets waits for ControlSockets.
```
this is not being used anywhere

amolgautam25 replied:
I have removed it.

fosterseth commented on tests/functional/mesh/reload_test.go:
```diff
@@ -0,0 +1,266 @@
+			"params":   "-c \"for i in {1..5}; do echo $i; sleep 3;done\"",
+		},
+	}
+	echoSleepShort := map[interface{}]interface{}{
```
your tests don't use echosleepshort or echosleeplong, so we can get rid of these definitions

fosterseth commented on tests/functional/mesh/reload_test.go:
```diff
@@ -0,0 +1,266 @@
+		"work-command": map[interface{}]interface{}{
+			"workType": "echosleeplong50",
+			"command":  "bash",
+			"params":   "-c \"for i in {1..5}; do echo $i; sleep 3;done\"",
```
this should count to 50

fosterseth commented on tests/functional/mesh/reload_test.go:
```diff
@@ -0,0 +1,266 @@
+		longCommand      map[interface{}]interface{}
+		extraLongCommand map[interface{}]interface{}
+	}{
+		{"normal_worker", echoSleepShort, echoSleepLong, echoSleepLong50},
```
we should probably have the kube_worker here, like in work_test.go

fosterseth commented on tests/functional/mesh/work_test.go:
```diff
@@ -371,7 +371,7 @@ func TestWork(t *testing.T) {
 				t.Fatal(err)
 			}
-			ctx, _ = context.WithTimeout(context.Background(), 10*time.Second)
+			ctx, _ = context.WithTimeout(context.Background(), 20*time.Second)
```
was this change helping with a flake?

fosterseth commented on tests/functional/mesh/reload_test.go:
```diff
@@ -0,0 +1,266 @@
+			modifiedData.Nodes = make(map[string]*mesh.YamlNode)
+			modifiedData.Nodes["node2"] = &mesh.YamlNode{
+				Connections: map[string]mesh.YamlConnection{},
+				Nodedef: []interface{}{
```
I don't think setting nodedef is necessary here

amolgautam25 replied:
I have made the changes

amolgautam25 committed:
Added Tests for live reload feature
Added helper functions for Live reload in `receptorcontrol.go`
Refactored `YamlNode`, Added `NodefConnections` and `NodedefBase`. Split `Nodedef` into `NodedefBase` and `NodefConnections`
Added `ModifyCLIMeshFromYaml` and `createNodedefConnections` in `climesh.go`
Modified existing MeshDefinitions in functional tests
Added `echoSleepLong50` work in `work_test.go`
Modified Makefile to output logs while running tests

amolgautam25 commented:
Still getting a ```--- FAIL: TestWork (0.00s)
    --- FAIL: TestWork/kube_worker/Reload_mesh_ (39.46s)
        work_test.go:664: write unix @->/tmp/receptor-testing/controlsockets/091534067/controlsock: write: broken pipe```

tried to do a ```close()``` and ```reconnect()``` but that causes another error.

fosterseth commented on tests/functional/mesh/work_test.go:
```diff
@@ -587,6 +607,128 @@ func TestWork(t *testing.T) {
+			}
+
+			// cancel the work so that it doesnt run after the test ends
+			_, err = controllers["node2"].WorkCancel(unitID)
```
let's assertworkcancel after we cancel

amolgautam25 replied:
I have made the required changes.

fosterseth commented on tests/functional/mesh/work_test.go:
```diff
@@ -587,6 +607,128 @@ func TestWork(t *testing.T) {
 		})
+		t.Run(testGroup+"/Reload mesh ", func(t *testing.T) {
+			t.Parallel()
+			controllers, m, _ := workSetup(t.Name())
```
add 
```
			if strings.Contains(t.Name(), "kube") {
				checkSkipKube(t)
			}
```

like we have in other tests

amolgautam25 replied:
I have made the changes , will push in next commit

fosterseth commented on tests/functional/mesh/work_test.go:
```diff
@@ -587,6 +607,128 @@ func TestWork(t *testing.T) {
 				t.Fatal(err)
 			}
 		})
+		t.Run(testGroup+"/Reload mesh ", func(t *testing.T) {
```
let's call this test `reload backends while streaming work results`

amolgautam25 replied:
I have made the changes , will push in next commit

fosterseth commented on tests/functional/mesh/work_test.go:
```diff
@@ -587,6 +607,136 @@ func TestWork(t *testing.T) {
+				t.Fatal(err)
+			}
+
+			// it takes some time for the streaming to start again
```
instead of sleeping, we should poll and check the size periodically, similar to AssertWorkResults

amolgautam25 replied:
I have made the required changes.

fosterseth commented on tests/functional/mesh/work_test.go:
```diff
@@ -59,6 +66,16 @@ func TestWork(t *testing.T) {
+			"namespace":  "default",
+			"kubeconfig": filepath.Join(home, ".kube/config"),
+			"image":      "alpine",
+			"command":    "sh -c \"for i in `seq 1 100`; do echo $i; sleep 4;done\"",
```
this doesn't count to 50?

amolgautam25 replied:
I corrected it. I was using 100 to test out something .

fosterseth commented on tests/functional/mesh/work_test.go:
```diff
@@ -70,17 +87,19 @@ func TestWork(t *testing.T) {
+		testGroup        string
+		shortCommand     map[interface{}]interface{}
+		longCommand      map[interface{}]interface{}
+		extraLongCommand map[interface{}]interface{}
```
let's rename this to longCommand50, to be consistent

amolgautam25 replied:
I have changed that

fosterseth commented on tests/functional/mesh/work_test.go:
```diff
@@ -587,6 +607,136 @@ func TestWork(t *testing.T) {
+			}
+
+			if newWorkStatus.StdoutSize <= workStatus.StdoutSize {
+				t.Fatal("Error : Work size is not increasing")
```
let's remove "Error : " and just start with "Work.."

amolgautam25 replied:
I have made the changes.
